### PR TITLE
feat: add program and workflow to sync sig-auth project board

### LINKED ISSUE #111963: [sig-auth] New project board and use github actions for board sync
- [x] Create a new project for sig-auth
   - We're using the classic board today: https://github.com/orgs/kubernetes/projects/54 and want to migrate to use the new board
- [ ] Use github actions to automate board sync
   - xref: https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/hack/sync-project-board.sh

/sig auth

### PR DESCRIPTION
**Pull request by aramase**
Signed-off-by: Anish Ramasekar <anish.ramasekar@gmail.com>

- Adds a go program to sync the sig-auth project board
  - mutation with `addProjectV2ItemById` adds the issues and PRs in a `No Status` column if not already in the board
- Adds a GitHub workflow to run at 6PM PST everyday that'll sync the project board
  - The workflow can also be triggered manually (hence the `workflow_dispatch` in the workflow yml

This has been validated with my own [project](https://github.com/orgs/kubernetes/projects/119/views/2?layout=board) and actions run: https://github.com/aramase/sig-auth-tools/actions/runs/3853956718/jobs/6567449739

fixes https://github.com/kubernetes/kubernetes/issues/111963

### CODE REVIEW

**Comment by aramase** [comment]
/hold

Work in progress. I've validated this with my own project in the `kubernetes` org

File: scripts/sync-project-board.sh [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+##
+## REREQS:
+##
+## This script assumes there is a github PAT in the GITHUB_TOKEN env var
+## that was created with the following permissions:
+##   - repo (all)
+##   - read:org
+##   - user (all)
+##   - read:enterprise
+##   - project (all)
```
Comment (enj) [inline scripts/sync-project-board.sh]: This seems really broad...

File: scripts/sync-project-board.sh [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+set -o pipefail
+
+## DESCRIPTION:
+##
+## This script queries all repos in a given github org and adds the issues 
+## with label 'sig/auth' to a specified project board.
+##
+## REREQS:
+##
+## This script assumes there is a github PAT in the GITHUB_TOKEN env var
```
Comment (enj) [inline scripts/sync-project-board.sh]: Whose PAT does this use?

File: scripts/sync-project-board.sh [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                    pageInfo {
+                        hasNextPage
+                        endCursor
+                    }
+                }
+            }
+        }
+    }' -f org="${GH_ORG}")"
+
+repos="$(jq ".data.viewer.organization.repositories.nodes[].name" <<< "$repos_json" |  tr -d '"' )"
```
Comment (enj) [inline scripts/sync-project-board.sh]:
```suggestion
repos="$(jq -r ".data.viewer.organization.repositories.nodes[].name" <<< "$repos_json")"
```

File: scripts/sync-project-board.sh [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+            }
+        }
+    }' -f org="${GH_ORG}" -F number="$1" --jq '.data.organization.projectV2.id')"
+    echo "$project_id"
+}
+
+# Get project ID
+project_id=$( get_project_id_from_number "$PROJECT_NUMBER" )
+echo "project id for issues (number $PROJECT_NUMBER): ${project_id}"
+# TODO(aramase): add PR project ID
```
Comment (enj) [inline scripts/sync-project-board.sh]: We would use the same project for both issues and PRs right?

File: scripts/sync-project-board.sh [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+set -e
+set -u
+set -o pipefail
+
+## DESCRIPTION:
+##
+## This script queries all repos in a given github org and adds the issues 
+## with label 'sig/auth' to a specified project board.
```
Comment (enj) [inline scripts/sync-project-board.sh]: Meta question: should this be a simple `go` program that we just `go run`?  I feel like every time we write a bash script in Kube we quickly outgrow it...

Note that I do not mean anything fancy: just a single `main.go` file with a `main` func (and `go.mod`).

File: scripts/sync-project-board.sh (hunk 2 above)
Comment (aramase) [reply scripts/sync-project-board.sh]: We'll need to generate a PAT and use it. I've been testing it with my own PAT here: https://github.com/aramase/sig-auth-tools/actions/runs/3853956718/jobs/6567449739. We'll need to set the PAT as a secret with name `SYNC_PROJECTS_PAT` before the action is run.

File: scripts/sync-project-board.sh (hunk 1 above)
Comment (aramase) [reply scripts/sync-project-board.sh]: I've narrowed it down it the following that works:

```bash
- repo (all)
- read (org)
- project (all)
```

`project (all)` is this which we'll need to mutate the project to add new items

Screenshot 2023-01-06 at 1 06 29 AM

**Comment by aramase** [comment]
/retitle feat: add program and workflow to sync sig-auth project board

File: main.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+	ctx := context.Background()
+	// GITHUB_TOKEN is a personal access token with the following scopes:
+	// - repo (all)
+	// - read:org
+	// - project (all)
+	token := os.Getenv("GITHUB_TOKEN")
+	ts := oauth2.StaticTokenSource(
+		&oauth2.Token{AccessToken: token},
+	)
+	tc := oauth2.NewClient(context.Background(), ts)
```
Comment (enj) [inline main.go]:
```suggestion
	tc := oauth2.NewClient(ctx, ts)
```

File: main.go [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+	projectName = "SIG Auth"
+)
+
+type ghClient struct {
+	*github.Client
+	v4Client *githubql.Client
+}
+
+func main() {
+	ctx := context.Background()
```
Comment (enj) [inline main.go]: Give this a timeout.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+func (c *ghClient) getProjectID(ctx context.Context, org, name string) (githubql.ID, error) {
+	var query struct {
+		Organization struct {
+			ProjectV2 struct {
+				Nodes []struct {
+					ID    githubql.ID
+					Title githubql.String
+				}
```
Comment (enj) [inline main.go]: Do these need struct tags?

File: scripts/sync-project-board.sh (hunk 2 above)
Comment (enj) [reply scripts/sync-project-board.sh]: Are we able to use one of the new style access token with the fine grained authz?

**Review by enj — APPROVED** [review]
Minor stuff.

File: scripts/sync-project-board.sh (hunk 2 above)
Comment (aramase) [reply scripts/sync-project-board.sh]: Not at the moment! fine-grained PAT doesn't support GraphQL API (xref: https://github.blog/2022-10-18-introducing-fine-grained-personal-access-tokens-for-github/#coming-next)

**Pull request merged**